### ISSUE TITLE
do while Loop

### ISSUE THREAD

**Issue by mawa290669** [fesch/Structorizer.Desktop#830]
I miss the possibility of displaying a foot-controlled loop according to C ++, C # or Java rules. There is a foot-controlled loop in the Structorizer ("BIS loop"). However, this corresponds to the functionality of the repeat-until loop from e.g. B. Turbo Pascal. But I need a foot-controlled loop after the function of a "do ... while loop". Is this possible?

**Comment by codemanyak** [comment]
Just invert the condition. This will automatically be corrected on code export.
I think it is obvious that Structorizer cannot interpret the condition in two contrary ways. We decided for the Repeat Until logic for historical reasons. (It was more common with programming languages before the C influence began to spread wide.)

**Comment by mawa290669** [comment]
Hallo Hr. Gürtzig,
ich hoffe, ich darf Ihnen auf deutsch antworten. So kann ich Ihnen mein Anliegen besser erklären. Ihre Idee, die Logik einfach umzukehren ist ja ok und wäre auch eine mögliche Lösung. Für meine Schüler, die ja gerade erst beginnen, eine Programmiersprache (C#) zu erlernen, ist dies nicht ganz so einfach. Könnte man nicht noch eine alternative fußgesteuerte Schleife anbieten, die der entspricht, wie es in C# o. Java der Fall ist???

Mfg. Wagner

**Comment by codemanyak** [comment]
@mawa290669 
Hallo Herr Wagner,
ich verstehe Ihr Anliegen durchaus glasklar. Aber meine (etwas überspitzte) Meinung dazu ist: Wer nicht in der Lage ist, eine logische Aussage zu negieren, wird auch kaum je programmieren lernen.

If you don't mind I will continue my answer in English (for the benefit of the user community).
A Nassi-Shneiderman diagram offers only one symbol for a foot-controlled loop. It cannot mean two contrary things at the same time in the same functional tool. Its semantics has to be defined either this or that way. This decision was made long ago. We might dispute, of course, which semantics is more sound: To have a re-entry condition or an exit condition at the end? My personal opinion is that it is more sound to interpret a condition as a guard allowing or not to pass the control to the code part that follows - which means that a head-controlled loop has an *entry* condition (= guard) and a foot-controlled loop has an *exit* condition (= guard). Kernighan and Ritchie obviously wanted to spare a keyword, so they reused "while" for the foot-controlled loop, which has several disadvantages, the most obvious of which is that the `while` clause of a **do-while** loop may easily be mistaken for the `while` clause of a **while** loop if one doesn't adhere very strictly to certain style guides.
But let us put the philosophy aside here. I think if a pupil does not understand the negation of a logical assertion then he/she cannot be taught programming either. They should definitely be trained to transform one algorithmic structure into an equivalent other structure and to operate on logical expressions. They must and can learn the difference between "until" and "while".

There might be a possible solution for your dilemma, though: In analogy to the **EXIT** element where we had introduced four different configurable keywords to distinguish the scope and effect (or e.g. the two **Instruction** keywords to identify *input* and *output* commands), we might offer a possibility to specify two distinct (and in fact contrary) keywords for the **REPEAT** loop in order to distinguish REPEAT UNTIL semantics from DO WHILE semantics. (The prefix keyword you *may* specify for the REPEAT loop in the Parser Preferences by now instructs Structorizer to *ignore* it when found before the condition. For backword compatibility reasons, however, it must still be possible to omit any prefix keyword in the **REPEAT** element text (i.e. before the actual Boolean expression), in which case the condition would have to work as exit condition as before.

There are some objections, though:
1. The wording would have to changed radically: We used to name the loop type "REPEAT UNTIL" all the time in Structorizer. It would be utterly confusing to let it do the contrary via a keyword.
1. The effect of the new keyword would just be the same as the insertion of a `not` operator but costs way more efforts. So why not just stick to negating the condition itself? Why should we have two different mechanisms to achieve the same logical result?
1. A Nassi-Shneiderman diagram is to express an algorithm, not a C# program. The aim is to present a well-defined algorithmically complete set of elements, not a maximum similarity to a specific programming language.
1. Using Structorizer, pupils will have to learn that there are differing syntactic conventions for similar things, anyway. Think of assignment symbols, for instance. They will have to get used to it. That's the real world. They can learn by watching and comparing the code preview. Discuss it. Don't spare them own experiences.

Briefly: I see indeed a possible way but I am very reluctant to go there. Or to say it in Bartleby's words: I'd prefer not to.
